pkg/asset/manifests/aws/zones.go

```diff
@@ -0,0 +1,272 @@
+	zones := []string{}
+	switch pool {
+	case types.MachinePoolControlPlaneRoleName:
+		if len(zo.controlPlaneZones) == 0 && len(defConfig) > 0 {
+			zones = defConfig
+		} else if len(zo.controlPlaneZones) == 0 {
+			zones = defRegion
+		}
+		zo.controlPlaneZones.Insert(zones...)
+
+	case types.MachinePoolComputeRoleName:
+		if len(zo.computeZones) == 0 && len(defConfig) > 0 {
+			zones = defConfig
+		} else if len(zo.computeZones) == 0 {
+			zones = defRegion
+		}
+		zo.computeZones.Insert(zones...)
+	}
+	zo.AvailabilityZones.Insert(zones...)
+}
+
+// setZones is the entrypoint to create the CAPI NetworkSpec structures
+// for managed or BYO VPC deployments from install-config.yaml.
+func setZones(in *zonesInput) error {
+	if len(in.InstallConfig.Config.AWS.Subnets) > 0 {
+		if err := in.GatherSubnetsFromMetadata(); err != nil {
+			return fmt.Errorf("failed to get subnets from metadata: %w", err)
+		}
+		return setZonesBYOVPC(in)
+	}
+
+	if err := in.GatherZonesFromMetadata(); err != nil {
+		return fmt.Errorf("failed to get availability zones from metadata: %w", err)
+	}
+	return setZonesManagedVPC(in)
+}
+
+// setZonesBYOVPC creates the CAPI NetworkSpec.Subnets setting the
+// desired subnets from install-config.yaml in the BYO VPC deployment.
+func setZonesBYOVPC(in *zonesInput) error {
+	if in.Cluster == nil {
+		return fmt.Errorf("failed to get AWSCluster config")
+	}
+	in.Cluster.Spec.NetworkSpec.VPC = capa.VPCSpec{
+		ID: in.Subnets.vpc,
+	}
+
+	for _, subnet := range in.Subnets.privateSubnets {
+		in.Cluster.Spec.NetworkSpec.Subnets = append(in.Cluster.Spec.NetworkSpec.Subnets, capa.SubnetSpec{
+			ID:               subnet.ID,
+			CidrBlock:        subnet.CIDR,
+			AvailabilityZone: subnet.Zone.Name,
+			IsPublic:         subnet.Public,
+		})
+	}
+
+	for _, subnet := range in.Subnets.publicSubnets {
+		in.Cluster.Spec.NetworkSpec.Subnets = append(in.Cluster.Spec.NetworkSpec.Subnets, capa.SubnetSpec{
+			ID:               subnet.ID,
+			CidrBlock:        subnet.CIDR,
+			AvailabilityZone: subnet.Zone.Name,
+			IsPublic:         subnet.Public,
+		})
+	}
+
+	return nil
+}
+
+// setZonesManagedVPC creates the CAPI NetworkSpec.VPC and the NetworkSpec.Subnets,
+// setting the desired zones from install-config.yaml in the managed
+// VPC deployment, when specified, otherwise default zones are discovered
+// from AWS API.
+// The CIDR blocks are calculated leaving free blocks to allow future expansions,
+// in Day-2, when desired.
+func setZonesManagedVPC(in *zonesInput) error {
+	out, err := extractZonesFromInstallConfig(in)
+	if err != nil {
+		return fmt.Errorf("failed to get availability zones: %w", err)
+	}
+
+	isPublishingExternal := in.InstallConfig.Config.Publish == types.ExternalPublishingStrategy
+	mainCIDR := capiutils.CIDRFromInstallConfig(in.InstallConfig)
+	in.Cluster.Spec.NetworkSpec.VPC = capa.VPCSpec{
+		CidrBlock: mainCIDR.String(),
+	}
+
```

Comment:
```suggestion
    allZones := out.AvailabilityZones()
```
if you use the function as suggested. Then use `allZones` instead of `out.AvailabilityZones` in the following code.